### PR TITLE
New feature: Export DBML to Oracle version 19c

### PR DESCRIPTION
**Pull request by huyleminh01**
## Summary
* Add oracle exporter to transform DBML code to Oracle SQL dialect (version 19c)
* Add test files

## Issue
https://github.com/holistics/dbml/issues/440

## Lasting Changes (Technical)

(please list down: code changes/things that have wide-effect; new libraries/functions added that can be used by others; examples below)

* Added `class OracleExporter` to transform DBML code to Oracle SQL dialect

## Checklist

Please check directly on the box once each of these are done

- [ ] Documentation (if necessary)
- [x] Tests (integration test/unit test)
- [ ] Integration Tests Passed
- [ ] Code Review

### CODE REVIEW

File: packages/dbml-core/__tests__/exporter/oracle_exporter/input/special_schema.in.dbml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+Table orders {
+  id "integer GENERATED ALWAYS AS IDENTITY PRIMARY KEY"
+  status order_status
+  "note" nclob [default: '']
+  user_id "integer" [not null, note: 'Store user uuid']
+  created_at timestamp [default: `current_timestamp`]
+
+  indexes {
+    (id, user_id) [name: 'IDX_ORDER_USER', type: btree, note: 'Ignore index  type', unique]
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/__tests__/exporter/oracle_exporter/input/special_schema.in.dbml:37]: I saw that we do not support export index's notes, and Oracle does not either.
[COMMENT](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/COMMENT.html#GUID-65F447C4-6914-4823-9691-F15D52DB74D7)

File: packages/dbml-core/types/export/index.d.ts [hunk 2]
```diff
@@ -1,4 +1,6 @@
-declare function _export(str: string, format: 'dbml' | 'mysql' | 'postgres' | 'json' | 'mssql'): string;
+import { ExportFormatOptions } from "./ModelExporter";
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/types/export/index.d.ts]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/src/export/utils.js [hunk 3]
```diff
@@ -39,3 +38,52 @@ export function buildNewTableName (firstTable, secondTable, usedTableNames) {
...
+  switch (database) {
+    case 'postgress':
+      escapeSignature = '"';
+      break;
+    case 'mysql':
+      escapeSignature = '`';
+      break;
+    case 'oracle':
+      escapeSignature = '"';
+      break;
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/src/export/utils.js]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/types/export/ModelExporter.d.ts [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import Database, { NormalizedDatabase } from '../model_structure/database';
+
+export type ExportFormatOptions = 'dbml' | 'mysql' | 'postgres' | 'json' | 'mssql' | 'oracle';
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/types/export/ModelExporter.d.ts]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/src/export/utils.js [hunk 5]
```diff
@@ -39,3 +38,52 @@ export function buildNewTableName (firstTable, secondTable, usedTableNames) {
...
+/**
+ *
+ * @param {string} schemaName
+ * @param {string} firstTableName
+ * @param {string} secondTableName
+ * @param {Map<string, Set>} tableNameListToCheck
+ * @returns string
+ * @description This function is a clone version of the buildNewTableName, but without side effect - update the original usedTableNames
+ */
+export function buildUniqueTableName (schemaName, firstTableName, secondTableName, tableNameListToCheck) {
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/src/export/utils.js]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/src/export/OracleExporter.js [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+      return line;
+    });
+
+    return lines;
+  }
+
+  static getTableContents (tableIds, model) {
+    const tableContentArr = tableIds.map((tableId) => {
+      const fieldContents = OracleExporter.getFieldLines(tableId, model);
+      const compositePKs = OracleExporter.getCompositePKs(tableId, model);
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/src/export/OracleExporter.js]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/src/export/OracleExporter.js [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+  static exportTables (tableIds, model) {
+    const tableContentList = OracleExporter.getTableContents(tableIds, model);
+
+    const tableStrs = tableContentList.map((tableContent) => {
+      const content = [...tableContent.fieldContents, ...tableContent.compositePKs];
+      const table = model.tables[tableContent.tableId];
+      const schema = model.schemas[table.schemaId];
+
+      const tableName = this.buildTableNameWithSchema(model, schema, table);
+      const constentString = content.map(line => `  ${line}`).join(',\n');
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/src/export/OracleExporter.js]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/src/export/OracleExporter.js [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+
+      const tableSet = usedTableNameMap.get(schema.name);
+
+      if (!tableSet) {
+        usedTableNameMap.set(schema.name, new Set(table.name));
+        return;
+      }
+
+      tableSet.add(table.name);
+    });
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/src/export/OracleExporter.js]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/src/export/OracleExporter.js [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+      }
+
+      // many to many relationship
+      const firstTableFieldsMap = buildJunctionFields1(refEndpoint.fieldIds, model);
+      const secondTableFieldsMap = buildJunctionFields2(foreignEndpoint.fieldIds, model, firstTableFieldsMap);
+
+      const newTableName = buildUniqueTableName(refEndpointSchema, refEndpointTable.name, foreignEndpointTable.name, usedTableNameMap);
+      const tableNameSet = usedTableNameMap.get(refEndpointSchema);
+      if (!tableNameSet) {
+        usedTableNameMap.set(refEndpointSchema, new Set(newTableName));
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/src/export/OracleExporter.js]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

File: packages/dbml-core/src/export/OracleExporter.js [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+    return `${shouldPrintSchema(schema, model) ? `${escapeObjectName(schema.name, 'oracle')}.` : ''}${escapeObjectName(table.name, 'oracle')}`;
+  }
+
+  static exportSchema (schemaName) {
+    return `CREATE USER ${escapeObjectName(schemaName, 'oracle')}\n`
+      + 'NO AUTHENTICATION\n'
+      + 'DEFAULT TABLESPACE system\n'
+      + 'TEMPORARY TABLESPACE temp\n'
+      + 'QUOTA UNLIMITED ON system;\n';
+  }
```
[the remark this answers is not in the document]
Comment (huyleminh01) [reply packages/dbml-core/src/export/OracleExporter.js]: I fixed it in this commit https://github.com/holistics/dbml/pull/534/commits/aec84bf4f646db60389aee4fcd46d0c674d64f9e

**Review by nguyenalter — APPROVED** [review]

**Pull request merged**